src/hw/raspi4/startup.s

```diff
@@ -0,0 +1,97 @@
+.section ".text.boot"
+
+.globl _Reset
+_Reset:
+  // read cpu id, stop slave cores
+  mrs x1, mpidr_el1
+  and x1, x1, #3
+  cbz x1, 2f
+  // cpu id > 0, stop
+1:wfe
+  b 1b
+2:// cpu id == 0
+  
+  // set stack before our code
+  ldr x1, =stack_top
+  
+  // set up EL1
+  mrs x0, CurrentEL
+  and x0, x0, #12 // clear reserved bits
+  
+  // running at EL3?
+  cmp x0, #12
+  bne 5f
+  // should never be executed, just for completeness
+  mov x2, #0x5b1
+  msr scr_el3, x2
+  mov x2, #0x3c9
+  msr spsr_el3, x2
+  adr x2, 5f
+  msr elr_el3, x2
+  eret
+  
+  // running at EL2?
+5:cmp x0, #4
+  beq 5f
+  msr sp_el1, x1
+  // enable CNTP for EL1
+  mrs x0, cnthctl_el2
+  orr x0, x0, #3
+  msr cnthctl_el2, x0
+  msr cntvoff_el2, xzr
+  // enable AArch64 in EL1
+  mov x0, #(1 << 31) // AArch64
+  msr hcr_el2, x0
+  mrs x0, hcr_el2
+  // enable FP / SIMD
```

Comment:
Hang on, I thought you didn't want Neon?

One of the reasons we don't enable neon is that neon registers aren't saved and restored by the thread switching mechanism. So I would invert this to explicitly *disable* Neon.
(assuming that doesn't trip up some library code but in that case leave it as is, at least any compiled code will be fine)

Reply:
It is because we will access `FPSR` when we store / load the CPU context.
We would be trapped if we did not enable that.

I think maybe I will add more comments about that.